src/s3.zig

```diff
@@ -0,0 +1,572 @@
+
+                break :brk try std.fmt.allocPrint(
+                    bun.default_allocator,
+                    "AWS4-HMAC-SHA256 Credential={s}/{s}/{s}/{s}/aws4_request, SignedHeaders={s}, Signature={s}",
+                    .{ this.accessKeyId, amz_day, region, service_name, signedHeaders, bun.fmt.bytesToHex(signature[0..DIGESTED_HMAC_256_LEN], .lower) },
+                );
+            }
+            const canonical = try std.fmt.bufPrint(&tmp_buffer, "{s}\n{s}\n\nhost:{s}\nx-amz-content-sha256:{s}\nx-amz-date:{s}\n\n{s}\n{s}", .{ method_name, normalizedPath, host, aws_content_hash, amz_date, signedHeaders, aws_content_hash });
+
+            var sha_digest = std.mem.zeroes(bun.sha.SHA256.Digest);
+            bun.sha.SHA256.hash(canonical, &sha_digest, JSC.VirtualMachine.get().rareData().boringEngine());
+
+            const signValue = try std.fmt.bufPrint(&tmp_buffer, "AWS4-HMAC-SHA256\n{s}\n{s}/{s}/{s}/aws4_request\n{s}", .{ amz_date, amz_day, region, service_name, bun.fmt.bytesToHex(sha_digest[0..bun.sha.SHA256.digest], .lower) });
+
+            const signature = bun.hmac.generate(sigDateRegionServiceReq, signValue, .sha256, &hmac_sig_service) orelse return error.FailedToGenerateSignature;
+
+            break :brk try std.fmt.allocPrint(
+                bun.default_allocator,
+                "AWS4-HMAC-SHA256 Credential={s}/{s}/{s}/{s}/aws4_request, SignedHeaders={s}, Signature={s}",
+                .{ this.accessKeyId, amz_day, region, service_name, signedHeaders, bun.fmt.bytesToHex(signature[0..DIGESTED_HMAC_256_LEN], .lower) },
+            );
+        };
+        errdefer bun.default_allocator.free(authorization);
+
+        if (signQuery) {
+            defer bun.default_allocator.free(host);
+            defer bun.default_allocator.free(amz_date);
+
+            return SignResult{
+                .amz_date = "",
+                .host = "",
+                .authorization = "",
+                .url = authorization,
+                .headers = .{
+                    .{ .name = "x-amz-content-sha256", .value = "" },
+                    .{ .name = "x-amz-date", .value = "" },
+                    .{ .name = "Authorization", .value = "" },
+                    .{ .name = "Host", .value = "" },
+                },
+            };
+        }
+        return SignResult{
+            .amz_date = amz_date,
+            .host = host,
+            .authorization = authorization,
+            .url = try std.fmt.allocPrint(bun.default_allocator, "https://{s}{s}", .{ host, normalizedPath }),
+            .headers = .{
+                .{ .name = "x-amz-content-sha256", .value = aws_content_hash },
+                .{ .name = "x-amz-date", .value = amz_date },
+                .{ .name = "Authorization", .value = authorization[0..] },
+                .{ .name = "Host", .value = host },
+            },
+        };
+    }
+
+    pub const S3StatResult = union(enum) {
```

Comment:
feels like we should have a generic for this like OptionalResult

Reply:
not sure about that, failure is present on each result, and not_found don't fit uploads.